### PR TITLE
OWLS-69993: Scripts to perform rolling restart of a WebLogic domain or cluster.

### PR DESCRIPTION
**Pull request by ankedia**
OWLS-69993 - Added scripts to perform rolling restart of WebLogic domain or a cluster. Updated the doc and README files. There's also a minor change to the usage section 'startServer.sh' and 'stopServer.sh' script to fix a minor issue.

### CODE REVIEW

File: kubernetes/samples/scripts/domain-lifecycle/helper.sh [hunk 1]
```diff
@@ -458,6 +458,109 @@ function getReplicaCount {
...
+}
+
+#
+# Function to create patch json to update cluster restartVersion
+# $1 - Domain resource in json format
+# $2 - Name of cluster whose replicas will be patched
+# $3 - replica count
+# $4 - Return value containing patch json string
+#
+function createPatchJsonToUpdateRestartVersion {
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/helper.sh:535]: Consider renaming to `createPatchJsonToUpdateClusterRestartVersion`

File: kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    m) kubernetesCli="${OPTARG}"
+    ;;
+    v) verboseMode=true;
+    ;;
+    h) usage 0
+    ;;
+    *) usage 1
+    ;;
+  esac
+done
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh:73]: Command line parsing problem: 

When I tried the above, it did not fail for "foo.sh somerandomstring" or "foo.sh -c myc somerandom -d foo". It only failed for "foo.sh -somerandomstring". 

This problem probably occurs in many of the scripts that use getopts. 

I assume "getopts" silently skips args that don't start with a "-".

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 3]
```diff
@@ -114,3 +114,41 @@ namespace          domain   cluster    min  max  goal  current  ready
 weblogic-domain-1  domain1  cluster-1  0    4    2     2        2
 weblogic-domain-1  domain1  cluster-2  0    4    0     0        0
 ```
+
+### Scripts to perform a rolling restart of a WebLogic domain or cluster
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:118]: It'd be nice to complete the suite of life-cycle scripts by including one that changes the introspectVersion.  I think the details about its usage are documented in the lifecycle docs...

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 4]
```diff
@@ -114,3 +114,41 @@ namespace          domain   cluster    min  max  goal  current  ready
...
+[2021-03-24T04:03:27.669 UTC][INFO] Successfully patched replicas for cluster 'cluster-1'!
+```
+
+Use the following command to roll the WebLogic Cluster Servers with a specific `restartVersion`:
+```shell
+$ rollCluster.sh -r v2 -c cluster-1 -d domain1 -n weblogic-domain-1
+[2021-03-24T13:46:16.833 UTC][INFO] Patching restartVersion for cluster 'cluster-1' to 'v2'.
+domain.weblogic.oracle/domain1 patched
+[2021-03-24T13:46:16.975 UTC][INFO] Successfully patched replicas for cluster 'cluster-1'!
+```
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:154]: For pretty much all of the helper scripts (roll, start/stop, scale), it's helpful to watch the pods and their labels react after running teh script:   `kubectl -n MYNS get pods --watch=true --show-labels`.   Perhaps this doc can mention the command and teh need to manually 'ctrl-C' out.  We can even consider adding a '-w' command to every life-cycle script that runs the command for you ...

**Review by tbarnes-us — COMMENTED** [review]
Only minor comments. LGTM overall. I can go ahead and approve if you want to disregard my comments.

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 5]
```diff
@@ -114,3 +114,41 @@ namespace          domain   cluster    min  max  goal  current  ready
...
 weblogic-domain-1  domain1  cluster-2  0    4    0     0        0
 ```
+
+### Scripts to perform a rolling restart of a WebLogic domain or cluster
+
+The `rollDomain.sh` script can be used to perform a rolling restart of the WebLogic Server Pods in a domain managed by the operator. Similarly, `rollCluster.sh` script can be used to perform a rolling restart of the WebLogic Server Pods belonging to a WebLogic cluster in a domain managed by the operator.
+
+The `rollDomain.sh` script updates the value of `spec.restartVersion` attribute of the domain resource.  The operator will then do a rolling restart of the Server Pods in the WebLogic domain once the value of `spec.restartVersion` is updated. You can provide the new value for `spec.restartVersion` as a parameter to the script or the script will automatically generate a new value to trigger the rolling restart. See the script `usage` information by using the `-h` option.
+
+```shell
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:124]: Minor nit: I suspect the 'shell' view slightly confusingly highlights keywords like 'for' and '!' in the 'INFO' lines. Maybe don't use the 'shell' keyword?

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 6]
```diff
@@ -138,17 +138,48 @@ domain.weblogic.oracle/domain1 patched
...
+The `introspectDomain.sh` script can be used to explicitly initiate the introspection of a WebLogic domain managed by the operator.  This script updates the value of `spec.introspectVersion` attribute of the domain resource to explicitly initiate the introspection. You can provide the new value for `spec.introspectVersion` as a parameter to the script or the script will automatically generate a new value to trigger the introspection. See the script `usage` information by using the `-h` option.
+
+```
+$ introspectDomain.sh -d domain1 -n weblogic-domain-1
+[2021-03-24T21:37:55.989 UTC][INFO] Patching introspectVersion for domain 'domain1' to '1'.
+domain.weblogic.oracle/domain1 patched
+[2021-03-24T21:37:56.110 UTC][INFO] Successfully patched introspectVersion for domain 'domain1'!
+```
+
+Use the following command to roll the Server Pods in a WebLogic domain with a specific `introspectVersion`:
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/README.md]: _This  text is misleading because the pods may or may not roll after a re-introspect.  The life-cycle doc for introspectVersion discusses the details.  So maybe reword this to something like:_

Use the following command to rerun a domain's introspect job. The resulting behavior depends on your domain home source type and other factors, see [link to the life-cycle doc for introspect version] for details.

File: kubernetes/samples/scripts/domain-lifecycle/restartServer.sh [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+  cat << EOF
+
+  This script restarts a running WebLogic server in a domain by deleting the server pod.
+ 
+  Usage:
+ 
+    $(basename $0) -s myserver [-n mynamespace] [-d mydomainuid] [-m kubecli]
+  
+    -s <server_name>           : Server name parameter is required.
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/restartServer.sh:22]: The WebLogic server name (not the pod name). This parameter is required.

File: kubernetes/samples/scripts/domain-lifecycle/restartServer.sh [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+# Validate that specified server is either part of a cluster or is an independent managed server
+validateServerAndFindCluster "${domainUid}" "${domainNamespace}" "${serverName}" isValidServer clusterName isAdminServer
+if [ "${isValidServer}" != 'true' ]; then
+  printError "Server ${serverName} is not part of any cluster and it's not an independent managed server. Please make sure that server name specified is correct."
+  exit 1
+fi
+
+printInfo "Initiating restart of '${serverName}' by deleting server pod '${domainUid}-${serverName}'."
+result=$(${kubernetesCli} -n ${domainNamespace} delete pod ${domainUid}-${serverName} --ignore-not-found)
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/restartServer.sh:96]: If this command fails, what happens?  I suspect it's possible that it can put something in 'stdout', in which case thsi script would incorrectly print "Server restart succeeded".

File: kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh [hunk 9]
```diff
@@ -12,28 +12,31 @@ function usage() {
 
   cat << EOF
 
-  This script performs rolling restart of a WebLogic cluster server pods in a domain by updating the
-  value of 'spec.clusters[<cluster-name>].restartVersion' attribute of the domain resource. This change
-  will cause the operator to perform a rolling restart of the WebLogic cluster server pods.
+  This script initiates a rolling restart of a WebLogic cluster server pods in a domain by updating
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh:15]: of a --> of the

File: kubernetes/samples/scripts/domain-lifecycle/restartServer.sh [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+fi
+
+# Validate that specified server is either part of a cluster or is an independent managed server
+validateServerAndFindCluster "${domainUid}" "${domainNamespace}" "${serverName}" isValidServer clusterName isAdminServer
+if [ "${isValidServer}" != 'true' ]; then
+  printError "Server ${serverName} is not part of any cluster and it's not an independent managed server. Please make sure that server name specified is correct."
+  exit 1
+fi
+
+printInfo "Initiating restart of '${serverName}' by deleting server pod '${domainUid}-${serverName}'."
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/restartServer.sh:95]: I'm fairly sure that the 'domainUid-serverName' string needs to be converted to a DNS 1123 compliant string (all lower case, and underscores changed to dashes).  If you don't already have the function, you can copy it from `./kubernetes/internal/utility.sh`.

File: kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh [hunk 11]
```diff
@@ -12,28 +12,31 @@ function usage() {
...
+  the value of 'spec.clusters[<cluster-name>].restartVersion' attribute of the domain resource.
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh:16]: the value of -> the value of the

File: kubernetes/samples/scripts/domain-lifecycle/rollDomain.sh [hunk 12]
```diff
@@ -12,9 +12,8 @@ function usage() {
 
   cat << EOF
 
-  This script performs rolling restart of pods in a WebLogic domain by updating the
-  value of 'spec.restartVersion' attribute of the domain resource. 
-  This change will cause the operator to perform a rolling restart of pods in WebLogic domain.
+  This script initiates a rolling restart of pods in a WebLogic domain by updating
+  the value of 'spec.restartVersion' attribute of the domain resource.
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/rollDomain.sh:16]: the value of -> the value of the

File: kubernetes/samples/scripts/domain-lifecycle/README.md (hunk 3 above)
Comment (tbarnes-us) [reply kubernetes/samples/scripts/domain-lifecycle/README.md]: Very nice! That was quick!

File: kubernetes/samples/scripts/domain-lifecycle/restartServer.sh (hunk 8 above)
Comment (ankedia) [reply kubernetes/samples/scripts/domain-lifecycle/restartServer.sh]: I have `set -eu` in the beginning so the script should abort if this command fails. I tried temporarily setting `kubernetesCli` value to "fake_kubectl" (before this command) to make it fail and the script failed with the below error. I can check the command return value if you think that's necessary, I thought it was not needed.

$ restartServer.sh -s managed-server3 -d domain1 -n weblogic-domain-1
[2021-03-25T02:02:11.430 UTC][INFO] Initiating restart of 'managed-server3' by deleting server pod 'domain1-managed-server3'.
restartServer.sh: line 98: fake_kubectl: command not found

File: kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+    -d <domain_uid>     : Domain unique-id. Default is 'sample-domain1'.
+
+    -n <namespace>      : Domain namespace. Default is 'sample-domain1-ns'.
+
+    -r <restartVersion> : Restart version. If this parameter is not provided, then the script will generate the
+                          'restartVersion' value of the cluster by incrementing the existing value. If the
+                          'restartVersion' value doesn't exist for the cluster then it will increment the domain
+                          'restartVersion'. If the domain 'restartVersion' also doesn't exist or the effective
+                          value is non-numeric, then the script will set the 'restartVersion' value to '1'.
```
[the remark this answers is not in the document]
Comment (ankedia) [reply kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh]: Yes. I tried this and it works.

File: kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh (hunk 2 above)
Comment (ankedia) [reply kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh:76]: Yes. There are many scripts that use "getopts" (including create-domain.sh and others) and have similar behavior. I'm not sure if it's critical to fail when args that don't start with "-" is specified since those args will be ignored.

File: docs-source/content/samples/simple/domains/lifecycle/_index.md [hunk 14]
```diff
@@ -7,7 +7,7 @@ description: "Start and stop Managed Servers, clusters, and domains."
 
 #### Domain lifecycle sample scripts
 
-Beginning in version 3.1.0, the operator provides sample scripts to start up or shut down a specific Managed Server or cluster in a deployed domain, or the entire deployed domain.
+Beginning in version 3.1.0, the operator provides sample scripts to start up or shut down a specific Managed Server or cluster in a deployed domain, or the entire deployed domain. Beginning in version 3.2.0, additional scripts are provided for scaling a WebLogic cluster, displaying the WebLogic cluster status, initiating rolling restart of a domain or a WebLogic cluster and initiating explicit introspection of a domain.
```
Comment (rjeberhard) [inline docs-source/content/samples/simple/domains/lifecycle/_index.md:10]: "cluster and" -> "cluster, and" (add comma)

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 15]
```diff
@@ -10,15 +10,15 @@ These scripts can be helpful when scripting the life cycle of a WebLogic Server
 The `startServer.sh` script starts a WebLogic Server in a domain. For clustered Managed Servers, either it increases the `spec.clusters[<cluster-name>].replicas` value for the Managed Server's cluster by `1` or updates the `spec.managedServers[<server-name>].serverStartPolicy` attribute of the domain resource or both as necessary. For the Administration Server, it updates the value of the `spec.adminServer.serverStartPolicy` attribute of the domain resource. For non-clustered Managed Servers, it updates the `spec.managedServers[<server-name>].serverStartPolicy` attribute of the domain resource. The script provides an option to keep the `spec.clusters[<cluster-name>].replicas` value constant for clustered servers. See the script `usage` information by using the `-h` option.
 
 Use the following command to start the server either by increasing the replica count or by updating the server start policy:
-```shell
+```
```
Comment (rjeberhard) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:13]: Why did you remove "shell"?

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 16]
```diff
@@ -114,3 +114,73 @@ namespace          domain   cluster    min  max  goal  current  ready
 weblogic-domain-1  domain1  cluster-1  0    4    2     2        2
 weblogic-domain-1  domain1  cluster-2  0    4    0     0        0
 ```
+
+### Scripts to initiate a rolling restart of a WebLogic domain or cluster
+
+The `rollDomain.sh` script can be used to initiate a rolling restart of the WebLogic Server Pods in a domain managed by the operator. Similarly, `rollCluster.sh` script can be used to initiate a rolling restart of the WebLogic Server Pods belonging to a WebLogic cluster in a domain managed by the operator.
```
Comment (rosemarymarano) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:120]: Similarly, `rollCluster.sh` script -> Similarly, the `rollCluster.sh` script

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 17]
```diff
@@ -114,3 +114,73 @@ namespace          domain   cluster    min  max  goal  current  ready
...
+
+The `rollDomain.sh` script updates the value of `spec.restartVersion` attribute of the domain resource.  The operator will then do a rolling restart of the Server Pods in the WebLogic domain once the value of `spec.restartVersion` is updated. You can provide the new value for `spec.restartVersion` as a parameter to the script or the script will automatically generate a new value to trigger the rolling restart. See the script `usage` information by using the `-h` option.
```
Comment (rosemarymarano) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:122]: value of `spec.restartVersion` attribute -> value of the `spec.restartVersion` attribute
The operator will then -> Then, the operator will 
once the value -> after the value

File: kubernetes/samples/scripts/domain-lifecycle/helper.sh [hunk 18]
```diff
@@ -458,6 +458,143 @@ function getReplicaCount {
...
+   __introspectVersion=0
+  fi
+  __introspectVersion=$((__introspectVersion+1))
+  eval $__result=${__introspectVersion}
+}
+
+#
+# Function to generate the cluster restartVersion by incrementing the
+# existing value. If the restartVersion doesn't exist at the cluster level, 
+# then it increments the domain level restartVersion value. If the effective
```
Comment (doxiao) [inline kubernetes/samples/scripts/domain-lifecycle/helper.sh:506]: It is not clear to me if the script actually increments the domain level restartVersion or just set the cluster restartVersion with the incremented domain level restartVersion.  I think it is latter, but not sure. If so, the wording needs to be clearer.

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 19]
```diff
@@ -114,3 +114,73 @@ namespace          domain   cluster    min  max  goal  current  ready
...
+
+Use the following command to roll the Server Pods in a WebLogic domain with a specific `restartVersion`:
+```
+$ rollDomain.sh -r v1 -d domain1 -n weblogic-domain-1
+[2021-03-24T13:43:47.586 UTC][INFO] Patching restartVersion for domain 'domain1' to 'v1'.
+domain.weblogic.oracle/domain1 patched
+[2021-03-24T13:43:47.708 UTC][INFO] Successfully patched restartVersion for domain 'domain1'!
+```
+
+The `rollCluster.sh` script updates the value of `spec.clusters[<cluster-name>].restartVersion` attribute of the domain resource. The operator will then do a rolling restart of the WebLogic cluster Server Pods once the value of `spec.clusters[<cluster-name>].restartVersion` is updated. You can provide the new value of `restartVersion` as a parameter to the script or the script will automatically generate a new value to trigger the rolling restart. See the script `usage` information by using the `-h` option.
```
Comment (rosemarymarano) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:139]: value of `spec.clusters -> value of the `spec.clusters
The operator will then -> Then, the operator will
once the value -> after the value

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 20]
```diff
@@ -114,3 +114,73 @@ namespace          domain   cluster    min  max  goal  current  ready
...
+```
+$ restartServer.sh -s managed-server1 -d domain1 -n weblogic-domain-1
+[2021-03-24T22:20:22.498 UTC][INFO] Initiating restart of 'managed-server1' by deleting server pod 'domain1-managed-server1'.
+[2021-03-24T22:20:37.614 UTC][INFO] Server restart succeeded !
+```
+
+### Scripts to explicitly initiate introspection of a WebLogic domain
+
+The `introspectDomain.sh` script can be used to rerun a WebLogic domain's introspect job by explicitly initiating the introspection. This script updates the value of `spec.introspectVersion` attribute of the domain resource. The resulting behavior depends on your domain home source type and other factors, see [Initiating introspection](https://oracle.github.io/weblogic-kubernetes-operator/userguide/managing-domains/domain-lifecycle/introspection/#initiating-introspection) for details. You can provide the new value of `introspectVersion` as a parameter to the script or the script will automatically generate a new value to trigger the introspection. See the script `usage` information by using the `-h` option.
```
Comment (rosemarymarano) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:166]: value of -> value of the

File: kubernetes/samples/scripts/domain-lifecycle/introspectDomain.sh [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+    $(basename $0) [-n mynamespace] [-d mydomainuid] [-i introspectVersion] [-m kubecli]
+  
+    -d <domain_uid>        : Domain unique-id. Default is 'sample-domain1'.
+
+    -n <namespace>         : Domain namespace. Default is 'sample-domain1-ns'.
+
+    -i <introspectVersion> : Introspect version. If this parameter is not provided, 
+                             then the script will generate the 'introspectVersion' by
+                             incrementing the existing value. If the 'spec.introspectVersion' 
+                             doesn't exist or it's value is non-numeric, then the script
```
Comment (doxiao) [inline kubernetes/samples/scripts/domain-lifecycle/introspectDomain.sh:29]: it's -> its

File: kubernetes/samples/scripts/domain-lifecycle/README.md [hunk 22]
```diff
@@ -114,3 +114,73 @@ namespace          domain   cluster    min  max  goal  current  ready
...
+```
+$ introspectDomain.sh -i v1 -d domain1 -n weblogic-domain-1
+[2021-03-24T21:38:34.369 UTC][INFO] Patching introspectVersion for domain 'domain1' to 'v1'.
+domain.weblogic.oracle/domain1 patched
+[2021-03-24T21:38:34.488 UTC][INFO] Successfully patched introspectVersion for domain 'domain1'!
+```
+
+### Watching the Pods after executing life cycle scripts
+
+After executing the various life cycle scripts described above for a domain or a cluster or a Server, you can manually run the `kubectl -n MYNS get pods --watch=true --show-labels` command to watch the effects of running the script and monitor the status and labels of various Pods. You will need to run 'Ctrl-C' in the end to stop watching the Pods and exit.
```
Comment (rosemarymarano) [inline kubernetes/samples/scripts/domain-lifecycle/README.md:186]: the various life cycle scripts -> the lifecycle scripts (life cycle (n), lifecycle (adj) 
to watch the effects -> to watch the effect
in the end to stop -> to stop

**Review by rosemarymarano — COMMENTED** [review]
Edits.

File: kubernetes/samples/scripts/domain-lifecycle/helper.sh (hunk 18 above)
Comment (doxiao) [reply kubernetes/samples/scripts/domain-lifecycle/helper.sh:506]: It is probably better in this case not to take the incremented domain level restartVersion, instead just set the cluster level restartVersion to 1.

File: kubernetes/samples/scripts/domain-lifecycle/rollDomain.sh [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+    $(basename $0) [-n mynamespace] [-d mydomainuid] [-r restartVersion] [-m kubecli]
+  
+    -d <domain_uid>     : Domain unique-id. Default is 'sample-domain1'.
+
+    -n <namespace>      : Domain namespace. Default is 'sample-domain1-ns'.
+
+    -r <restartVersion> : Restart version. If this parameter is not provided, 
+                          then the script will generate the 'restartVersion' by
+                          incrementing the existing value. If the 'spec.restartVersion' 
+                          doesn't exist or it's value is non-numeric, then the script
```
Comment (doxiao) [inline kubernetes/samples/scripts/domain-lifecycle/rollDomain.sh:29]: it's -> its

**Review by doxiao — COMMENTED** [review]
LGTM. Just had a couple of minor comments.

File: kubernetes/samples/scripts/domain-lifecycle/README.md (hunk 15 above)
Comment (ankedia) [reply kubernetes/samples/scripts/domain-lifecycle/README.md:13]: Tom mentioned that the "shell" view was confusingly highlighting keywords like 'for' and '!' in the 'INFO' lines. I didn't see much difference in the formatting when not using "shell".  Do you see any issue or difference in formatting due to this change? Thanks.

File: kubernetes/samples/scripts/domain-lifecycle/helper.sh (hunk 18 above)
Comment (ankedia) [reply kubernetes/samples/scripts/domain-lifecycle/helper.sh]: I'm not sure if the cluster will roll if the domain level restartVersion is 1 and then the cluster level restartVersion is also updated to 1.

File: kubernetes/samples/scripts/domain-lifecycle/README.md (hunk 15 above)
Comment (rjeberhard) [reply kubernetes/samples/scripts/domain-lifecycle/README.md:13]: I had just gone through and added `shell` because it seemed to do better formatting than `bash`. I thought it was helpful :).

Comment (rjeberhard) [reply kubernetes/samples/scripts/domain-lifecycle/README.md:13]: Okay, we can leave this alone now, but I'll probably come through later and restore the `shell`, but have the output be in a separate block.

**Review by rjeberhard — APPROVED** [review]

File: kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh (hunk 2 above)
Comment (tbarnes-us) [reply kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh:76]: Ok.

File: kubernetes/samples/scripts/domain-lifecycle/restartServer.sh (hunk 8 above)
Comment (tbarnes-us) [reply kubernetes/samples/scripts/domain-lifecycle/restartServer.sh]: I forgot about teh 'set -eu'.

File: kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+if [ -z "${restartVersion}" ]; then
+  generateClusterRestartVersion "${domainJson}" "${clusterName}" restartVersion 
+fi
+
+printInfo "Patching restartVersion for cluster '${clusterName}' to '${restartVersion}'."
+createPatchJsonToUpdateClusterRestartVersion "${domainJson}" "${clusterName}" "${restartVersion}" patchJson
+
+executePatchCommand "${kubernetesCli}" "${domainUid}" "${domainNamespace}" "${patchJson}" "${verboseMode}"
+
+printInfo "Successfully patched replicas for cluster '${clusterName}'!"
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/domain-lifecycle/rollCluster.sh:123]: replicas -> restartVersion

**Review by anpanigr — COMMENTED** [review]
Update the PR with Jenkin Result.

**Review by tbarnes-us — APPROVED** [review]

**Review by doxiao — APPROVED** [review]
